### ISSUE TITLE
[GPU][CUDA] Add support to linear tree with device=gpu, device=cuda

### ISSUE THREAD

**Issue by dragonbra** [microsoft/LightGBM#6555]
## Summary

Add support to linear tree with `device=gpu`, `device=cuda`

## Motivation

Linear tree is a useful feature to improve model's performance. However, it can be only used in CPU version now. I think it would be great if we can add support in GPU and CUDA version

## Description

By reading the code, I found that the `gpu_tree_learner` now calculates the gradients and hessians in GPU device and pass them back to CPU to call the `train()` function in `serial_tree_learner`. The `linear_tree_learner`'s training process is based on `serial_tree_learner`'s and used Eigen library, so it should be easy to implement this feature if we do not take GPU acceleration of training linear model leaves into consideration. 
I have already successfully tried it in `device=gpu` version and I will submit a PR when I have time to organize the code.
I haven't done experiments in `device=cuda` version. It would be great if anyone can contribute to it.

## References

Trees with linear models at leaves #3299 
https://github.com/microsoft/LightGBM/blob/f8ec57b8eba69b7371324bba1d158f850802c782/src/treelearner/linear_tree_learner.cpp#L62
